### ISSUE TITLE
Skaffold not working with namespace manifests, skaffold dev inconsistent deployment/cleanup.

### ISSUE THREAD

**Issue by ZeroVocabulary** [GoogleContainerTools/skaffold#6285]
### Expected behavior
When providing a namespace manifest (kubectl), and mentioning that namespace in skaffold.yaml (deploy.kubectl.defaultNamespace), skaffold should create the namespace and deploy the rest of the kubectl manifests.
Namespace manifest below:
```
apiVersion: v1
kind: Namespace
metadata:
  name: <insert-namespace-name-here>
```

### Actual behavior
Skaffold does not properly use namespace manifests, however it does try to:
```
Starting deploy...
 - namespace/test created
 - Error from server (NotFound): error when creating "STDIN": namespaces "test" not found
 - Error from server (NotFound): error when creating "STDIN": namespaces "test" not found
kubectl apply: exit status 1
```

When using skaffold deploy and skaffold delete afterwards, it fails 10/10 times. Powershell:
`1..10 | % {echo "RUN $_ ================================================================"; skaffold deploy; skaffold delete}
`

When creating the namespace before skaffold deploy, it succeeds 10/10 times. Powershell:
`1..10 | % {echo "RUN $_ ================================================================"; kubectl create namespace test; skaffold deploy; skaffold delete}
`

However, when using skaffold dev, upon failure, it does NOT delete the namespace. It creates the namespace but then exits when it gets the namespace not found error, without deleting the namespace.
```
Starting deploy...
 - namespace/test created
 - Error from server (NotFound): error when creating "STDIN": namespaces "test" not found
 - Error from server (NotFound): error when creating "STDIN": namespaces "test" not found
kubectl apply: exit status 1
```

So skaffold dev alternates between succeeding and failing, because the namespace will exist after failing, leading to a successful run the next time, and after a successful run, it will delete the namespace, leading to a failure the next time.

When doing skaffold delete and creating the namespace before skaffold dev, it succeeds 10/10 times.
`skaffold delete; kubectl create namespace test; skaffold dev;
`

### Information

- Skaffold version: v1.28.1
- Operating system: Microsoft Windows 10 Home 10.0.19043 
- Installed via: skaffold.dev 
- Contents of skaffold.yaml:

```yaml
apiVersion: skaffold/v2beta19
kind: Config
build:
  artifacts: []
deploy:
  kubectl:
    manifests:
      - kubernetes-configs\*
    defaultNamespace: test
```

### Steps to reproduce the behavior

1. https://github.com/ZeroVocabulary/skaffold-namespace-manifests-inconsistency
2. Run `skaffold dev` repeatedly, it will alternate between successful deployment and failure.

**Comment by ZeroVocabulary** [comment]
## Design suggestions

I can't tell whether support for namespace manifests is intentional or accidental.

### If namespace manifests are supported
It would be nice to be able to use the namespace within that skaffold deployment (right now skaffold just tells you it does not exist, even though you instructed skaffold to create it). As far as I know there isn't any way to create a namespace if it does not exist for kubectl. It is possible with helm, because of the helm cli option:
[deploy-helm-releases-createNamespace](https://skaffold.dev/docs/references/yaml/#deploy-helm-releases-createNamespace
)

If that is not possible, skaffold dev should at least delete it on failure. In my experience with skaffold, I've never seen it only partially delete things on exit. These namespace manifests are the only time (maybe I'm wrong though).

If namespace manifests are supported, namespace manifests can cause confusion/damage because skaffold will delete the namespace and its contents afterward, even if they are not part of the deployment. The namespace may have existed before the deployment, and its contents may have existed before the deployment, but they will all get deleted. Some solutions:
- Maybe it could delete only if the namespace is empty.
- Or it could never delete (skip namespace manifests during cleanup).
- Add an option for it.
- Keep it the way it is. If the user includes a namespace manifest, they are expected to know that it will be deleted on cleanup like anything else they include.


I like to use skaffold for my integration test system, where the namespace is a randomly generated name, and I want it deleted at the end. So it would be unfortunate for me if deletion were not possible; I would need to use kubectl delete after using skaffold.

### If namespace manifests are not supported
They should be skipped, not deployed.

**Comment by ZeroVocabulary** [comment]
I found out how to do namespace manifests. You add them first, before everything else:
```
deploy:
  kubectl:
    manifests:
      - kubernetes-configs\namespaces\*
      - kubernetes-configs\*
```
The namespace will be deleted when using skaffold delete, and will not have issues with skaffold dev (the namespace will be deleted properly).

**Comment by ZeroVocabulary** [comment]
I thought skaffold dev always cleaned up properly, and namespace manifests were causing an issue, but I did the same without namespace manifests:
https://github.com/ZeroVocabulary/skaffold-namespace-manifests-inconsistency/tree/main/test1

There are 2 deployments being made, "nginx" in a namespace that does not exist, and "nginx2" in the default namespace (it does exist). When using skaffold dev, it fails, and "nginx2" is not deleted. I thought this shouldn't happen, which is why I thought it was a bug, but now I'm not sure. There is no namespace manifest here and it still happens. Is this intended?

If skaffold dev is not meant to always delete on failure, then it is not a bug (that would be pretty confusing though, skaffold deploy;skaffold delete always works).

If skaffold dev is meant to delete everything on failure, then its a bug, and it is caused by trying to use namespaces that do not exist, not by the namespace manifests as I previously claimed.

The namespace manifests not getting priority is awkward but can be bypassed as explained in this [comment](https://github.com/GoogleContainerTools/skaffold/issues/6285#issuecomment-890326634).
  
Should namespaces get priority? When using kubectl apply to deploy all the manifests at once (namespace manifest and manifests that use that namespace), it gives the same error skaffold gives:
```
deployment.apps/nginx2 created
namespace/test created
Error from server (NotFound): error when creating "kubernetes-configs\\nginx_Deployment.yaml": namespaces "test" not found
```
In my opinion, namespaces shouldn't get priority because it would cause differences between kubectl apply and skaffold.

But skaffold dev should still delete everything imo (like skaffold deploy;skaffold delete would), its confusing that it doesn't, and feels like a bug.

**Comment by nkubala** [comment]
@ZeroVocabulary thanks for the issue! there are two things going on here that I see.

1) providing both a manifest that creates a namespace for you (we'll call it `namespace.yaml`) and then attempting to use the created namespace in a subsequent deployment (your nginx deployment, let's call it `nginx.yaml`) is a totally valid use case, but one that creates effectively a race in skaffold. skaffold needs that namespace to exist when it sends `nginx.yaml` to `kubectl apply`, and in practice this never happens, so the deployment fails every time.

2) when this deploy in skaffold fails, the created namespace should get cleaned up, but it doesn't. this is a bug! I think I found the root cause and addressed it in #6345.

for the first point, the way your project is structured won't support the use case you're trying to accomplish. there's no way to tell skaffold "apply manifest 1, **wait for it to be ready**, and then apply manifest 2". but you're not totally out of luck! using skaffold's `requires` clause to specify config dependencies, we can do this:

```
➜  tree
.
├── deployment.yaml <- contains a reference to the namespace you want to create
└── skaffold.yaml
├── namespace
│   ├── namespace.yaml <- contains the definition for your namespace
│   └── skaffold.yaml
```

your top-level `skaffold.yaml` looks like:
```yaml
apiVersion: skaffold/v2beta19
kind: Config
requires:
- path: ../namespace
deploy:
  kubectl:
    manifests:
    - deployment.yaml
    defaultNamespace: test
```

and the `namespace/skaffold.yaml` looks like:
```
apiVersion: skaffold/v2beta19
kind: Config
deploy:
  kubectl:
    manifests:
      - namespace.yaml
```

now, just running `skaffold dev` from the root gives you the behavior you're looking for - specifying that the top-level skaffold deployment `requires` the namespaces deployment causes skaffold to wait for it to be ready before moving on to the next deployment.

```
➜  skaffold dev
Listing files to watch...
Generating tags...
Checking cache...
Starting test...
Tags used in deployment:
Starting deploy...
 - namespace/test configured
 - deployment.apps/nginx created
Waiting for deployments to stabilize...
 - test:deployment/nginx is ready.
Deployments stabilized in 2.378 seconds
Press Ctrl+C to exit
Watching for changes...
```

hopefully this is what you were looking for! let me know if there are other enhancements you think skaffold should make here, otherwise feel free to close this issue.

**Comment by ZeroVocabulary** [comment]
Yes I used that solution, its in one of the above comments. I didn't need multiple files though, I just did it like this:
```
deploy:
  kubectl:
    manifests:
      - kubernetes-configs\namespaces\*
      - kubernetes-configs\*
```
And the skaffold dev cleanup bug isn't actually related to namespaces, it happens whenever the kubectl deployment fails.
I'll close this issue, thank you!
